microdnf-plugins/changelog_plugin/changelog_cmd_plugin.cpp

```diff
@@ -0,0 +1,94 @@
+#include "changelog.hpp"
+
+#include <microdnf/iplugin.hpp>
+
+#include <iostream>
+
+using namespace microdnf;
+
+namespace {
+
+constexpr const char * PLUGIN_NAME = "dnf5_test_plugin";
+constexpr Version PLUGIN_VERSION{0, 1, 0};
+constexpr const char * PLUGIN_DESCRIPTION = "Test plugin.";
+constexpr const char * PLUGIN_AUTHOR_NAME = "Jaroslav Rohel";
+constexpr const char * PLUGIN_AUTHOR_EMAIL = "jrohel@redhat.com";
+
+class ChangelogCmdPlugin : public IPlugin {
+public:
+    Version get_api_version() const noexcept override { return PLUGIN_API_VERSION; }
+
+    const char * get_name() const noexcept override { return PLUGIN_NAME; }
+
+    Version get_version() const noexcept override { return PLUGIN_VERSION; }
+
+    const char * get_attribute(const char * attribute) const noexcept override {
+        if (std::strcmp(attribute, "author_name") == 0)
+            return PLUGIN_AUTHOR_NAME;
+        if (std::strcmp(attribute, "author_email") == 0)
+            return PLUGIN_AUTHOR_EMAIL;
+        if (std::strcmp(attribute, "description") == 0)
+            return PLUGIN_DESCRIPTION;
+        return nullptr;
+    }
+
+    void init(microdnf::Context * context) override { this->context = context; }
+
+    std::vector<std::unique_ptr<libdnf::cli::session::Command>> create_commands(
+        libdnf::cli::session::Command & parent) override;
+
+    //bool hook(HookId) override { return true; }
+
+    void finish() noexcept override {}
+
+private:
+    microdnf::Context * context;
+};
+
+
+std::vector<std::unique_ptr<libdnf::cli::session::Command>> ChangelogCmdPlugin::create_commands(
+    libdnf::cli::session::Command & parent) {
+    std::vector<std::unique_ptr<libdnf::cli::session::Command>> commands;
+    commands.push_back(std::make_unique<ChangelogCommand>(parent));
+    return commands;
+}
+
+
+}  // namespace
+
+
+Version microdnf_plugin_get_api_version(void) {
+    return PLUGIN_API_VERSION;
+}
+
+const char * microdnf_plugin_get_name(void) {
+    return PLUGIN_NAME;
+}
+
+Version microdnf_plugin_get_version(void) {
+    return PLUGIN_VERSION;
+}
+
+IPlugin * microdnf_plugin_new_instance(Version api_version) try {
+    if (api_version.major != PLUGIN_API_VERSION.major || api_version.minor != PLUGIN_API_VERSION.minor) {
```

Comment:
I think the version check should be done in microdnf, not in each plugin (which is prone to inconsistencies).

Also, should we assign semantics to the `major` and `minor` version, `major` being backwards incompatible breakage and `minor` backwards compatible additions? Then the check would be that the plugin's `major` must be equal and plugin's `minor` must be less or equal.

Reply:
I'll add semantics and move control to microdnf. It will be less flexible but also easier to use.